Cleanup: Get rid of MIMEFieldWrapper from HPACK encoding

masaori335 opened the pull request:
With general cleanups
- Introduce `HpackHeaderField` for replacement of passing around set of `const char *name, int name_len, const char *value, int value_len`.
- ~~Remove `HpackIndexingTable &indexing_table` from "low level interface" of encoding for clarification.~~

This depends on #6509 to avoid conflicts.

maskit commented on proxy/http2/HPACK.cc:
```diff
@@ -877,38 +918,46 @@ hpack_encode_header_block(HpackIndexingTable &indexing_table, uint8_t *out_buf,
+
     const HpackLookupResult result = indexing_table.lookup(header);
+
+    if (result.match_type != HpackMatch::EXACT && field_type == HpackField::INDEXED_LITERAL) {
```
I'm not sure what was unclear, but the intent was probably to delegate whole actual work. Now you can encode headers without updating indexing table, which should not happen. The old interface (with indexing_table param) ensures such an accident never happens.

bryancall commented:
[approve ci]

maskit commented:
@masaori335 Can you explain why the new interface that looks unsafe to me is actually better?

masaori335 commented:
It simplifies low-level encoding functions, they can focus on encoding. In general, giving 2 tasks to a function ( encoding & updating the table in this case) is bad practice. But the change is trivial, so I moved out. ( Making 2 changes in a PR is bad practice as well:) )

Rebased on the latest master. Ready for review.

maskit commented:
> In general, giving 2 tasks to a function ( encoding & updating the table in this case) is bad practice. 

Then you should remove all the high level interfaces :p

I don't think the practice applies to the encode function because you cannot do only one of them, and there's nothing to do between the two. Having two separate functions implies there is a case you want to call only either of them, but if you do, it breaks HPACK state. That's why I think it should be a single atomic operation.

masaori335 commented:
The reason why we have these low-level APIs is just for unit tests. The high-level API has the responsibility for the state of HPACK. We don't need to give the responsibility to the low-level APIs. Whatever, it's out of scope from this PR now.

SolidWallOfCode commented:
Would it be possible to use `std::string_view` or `ts::TextView` instead of explicit `char const *` and `size_t`?

masaori335 commented:
@SolidWallOfCode Replaced with `std::string_view`

masaori335 commented on proxy/http2/HPACK.cc:
```diff
@@ -227,19 +223,15 @@ hpack_parse_field_type(uint8_t ftype)
-      if (ptr_len_casecmp(name, name_len, table_name, table_name_len) == 0) {
-        if ((value_len == table_value_len) && (memcmp(value, table_value, value_len) == 0)) {
+      if (strcasecmp(header.name, STATIC_TABLE[index].name) == 0) {
+        if (header.value.compare(STATIC_TABLE[index].value) == 0) {
```
@SolidWallOfCode `compare` is as fast as size check & `memcmp`, right?

masaori335 replied:
hmm, with llvm-11, maybe not
```
    _LIBCPP_CONSTEXPR_AFTER_CXX11 int compare(basic_string_view __sv) const _NOEXCEPT
    {
        size_type __rlen = _VSTD::min( size(), __sv.size());
        int __retval = _Traits::compare(data(), __sv.data(), __rlen);
        if ( __retval == 0 ) // first __rlen chars matched
            __retval = size() == __sv.size() ? 0 : ( size() < __sv.size() ? -1 : 1 );
        return __retval;
    }
```

masaori335 replied:
I restored to size check & `memcmp`

SolidWallOfCode replied:
You need to compile with -O2 to see what happens in production code. Checkout the results on godbolt.org - if you use this code
```
#include <string_view>

bool f(std::string_view const& lhs, std::string_view const& rhs) {
    return lhs == rhs;
}
```
and compile with -O0 then you get the calls to compare. If you change that to -O2, then with LLVM 11.0.1 you get this:
```
push    rax
mov     rdx, qword ptr [rdi]
cmp     rdx, qword ptr [rsi] ; check lengths
jne     .LBB0_5 ; return false if different.
test    rdx, rdx ; same pointer?
je      .LBB0_3 ; yep, return true.
mov     rdi, qword ptr [rdi + 8]
mov     rsi, qword ptr [rsi + 8]
call    memcmp ; compare the strings.
```

masaori335 replied:
I completely missed what -O2 does. Thanks for let me know, I'll keep in mind.

SolidWallOfCode commented on proxy/http2/HPACK.cc:
```diff
@@ -227,19 +223,18 @@ hpack_parse_field_type(uint8_t ftype)
-      if (ptr_len_casecmp(name, name_len, table_name, table_name_len) == 0) {
-        if ((value_len == table_value_len) && (memcmp(value, table_value, value_len) == 0)) {
+      if (strcasecmp(header.name, name) == 0) {
+        if (header.value.size() == value.size() && memcmp(header.value.data(), value.data(), value.size()) == 0) {
```
I think `memcmp` is overloaded the same way as `strcasecmp`, e.g. you could do
```
if (memcmp(header.value, value) == 0) {
```
However, the equality operator also uses `memcmp` so both of these are equivalent to
```
if (header.value == value) {
```

masaori335 replied:
I completely forgot about `memcmp` in `TextView.h`.  Will call it.

masaori335 replied:
It looks like the equality operator calls `compare` in some implementations (- e.g. libc++ ). `compare` seems bit slow, so I'll go with `memcmp` in `TextView.h`

masaori335 replied:
Done.

SolidWallOfCode commented on proxy/http2/HPACK.cc:
```diff
@@ -644,12 +621,11 @@ encode_literal_header_field_with_new_name(uint8_t *buf_start, const uint8_t *buf
+  int name_len              = header.name.size();
+  const char *original_name = header.name.data();
 
   ts::LocalBuffer<char> local_buffer(name_len);
```
How is `LocalBuffer<char> local_buffer(name_len);` better than `char local_buffer[name_len];`?

masaori335 replied:
In case of `name_len` is huge to allocate on thread stack. If `name_len` is small, it's equivalent.

SolidWallOfCode commented on proxy/http2/HPACK.cc:
```diff
@@ -946,22 +922,25 @@ hpack_encode_header_block(HpackIndexingTable &indexing_table, uint8_t *out_buf,
-        (ptr_len_casecmp(name, name_len, MIME_FIELD_AUTHORIZATION, MIME_LEN_AUTHORIZATION) == 0)) {
+    HpackField field_type;
+    // TODO: replace `strcasecmp` with `memcmp`
+    if ((strcasecmp(name, HPACK_HDR_FIELD_COOKIE) == 0 && value.size() < 20) ||
```
If you're tuning, do the size comparison first.

masaori335 replied:
Should be. Thanks!

masaori335 replied:
Done.

SolidWallOfCode commented on proxy/http2/HPACK.cc:
```diff
@@ -946,22 +922,25 @@ hpack_encode_header_block(HpackIndexingTable &indexing_table, uint8_t *out_buf,
 
   MIMEFieldIter field_iter;
   for (MIMEField *field = hdr->iter_get_first(&field_iter); field != nullptr; field = hdr->iter_get_next(&field_iter)) {
```
Why doesn't this use the STL paradigm of iterators with `begin` / `end`?

masaori335 replied:
Because `MIMEHdr` doesn't follow STL paradigm.

SolidWallOfCode replied:
That was actually my question - perhaps we should take `MIMEHdr` to do that.

masaori335 replied:
Filed an issue > https://github.com/apache/trafficserver/issues/7472

zwoop commented:
Cherry-picked to v9.1.x branch.
